sw/device/tests/dif/dif_pinmux_test.cc

```diff
@@ -0,0 +1,378 @@
+}
+
+namespace dif_pinmux_test {
+namespace {
+using testing::Test;
+using mock_mmio::MmioTest;
+
+class DifPinmuxTest : public Test, public MmioTest {
+ protected:
+  void DeviceInit() {
+    EXPECT_READ32(PINMUX_REGEN_REG_OFFSET, 1);
+
+    // Reset sequence.
+    std::vector<ptrdiff_t> insel_registers{
+        PINMUX_PERIPH_INSEL0_REG_OFFSET, PINMUX_PERIPH_INSEL1_REG_OFFSET,
+        PINMUX_PERIPH_INSEL2_REG_OFFSET, PINMUX_PERIPH_INSEL3_REG_OFFSET,
+        PINMUX_PERIPH_INSEL4_REG_OFFSET, PINMUX_PERIPH_INSEL5_REG_OFFSET,
+        PINMUX_PERIPH_INSEL6_REG_OFFSET,
+    };
+    for (const auto &reg : insel_registers) {
+      // TODO - when fixed, drop the explicit cast.
+      EXPECT_WRITE32(reg, (uint32_t)0);
+    }
+
+    std::vector<ptrdiff_t> outsel_registers{
+        PINMUX_MIO_OUTSEL0_REG_OFFSET, PINMUX_MIO_OUTSEL1_REG_OFFSET,
+        PINMUX_MIO_OUTSEL2_REG_OFFSET, PINMUX_MIO_OUTSEL3_REG_OFFSET,
+        PINMUX_MIO_OUTSEL4_REG_OFFSET, PINMUX_MIO_OUTSEL5_REG_OFFSET,
+        PINMUX_MIO_OUTSEL6_REG_OFFSET,
+    };
+    for (const auto &reg : outsel_registers) {
+      // TODO - autogenerate.
+      EXPECT_WRITE32(reg, 0x20820820);
+    }
+
+    mmio_region_t base_addr = dev().region();
+    EXPECT_EQ(dif_pinmux_init(base_addr, &dif_pinmux_),
+              kDifPinmuxInitErrorNone);
+
+    EXPECT_TRUE(dif_pinmux_.write_enabled);
+  }
+
+  dif_pinmux_t dif_pinmux_;
+};
+
+// dif_pinmux_init() tests.
+
+TEST_F(DifPinmuxTest, InitFailureInvalidArgumentPinmux) {
+  mmio_region_t base_addr = dev().region();
+  EXPECT_EQ(dif_pinmux_init(base_addr, nullptr),
+            kDifPinmuxInitErrorInvalidArgument);
+}
+
+TEST_F(DifPinmuxTest, InitFailureRegistersLocked) {
+  EXPECT_READ32(PINMUX_REGEN_REG_OFFSET, (uint32_t)0);
+
+  mmio_region_t base_addr = dev().region();
+  EXPECT_EQ(dif_pinmux_init(base_addr, &dif_pinmux_),
+            kDifPinmuxInitErrorRegistersLocked);
+
+  EXPECT_FALSE(dif_pinmux_.write_enabled);
+}
+
+TEST_F(DifPinmuxTest, InitSuccess) { DeviceInit(); }
+
+// dif_pinmux_registers_lock() tests.
+
+TEST_F(DifPinmuxTest, RegistersLockFailureInvalidArgumentPinmux) {
+  EXPECT_EQ(dif_pinmux_registers_lock(nullptr),
+            kDifPinmuxInitErrorInvalidArgument);
+}
+
+TEST_F(DifPinmuxTest, RegistersLockSuccess) {
+  DeviceInit();
+
+  // TODO - when fixed, drop the explicit cast.
+  EXPECT_WRITE32(PINMUX_REGEN_REG_OFFSET, (uint32_t)0);
+
+  EXPECT_EQ(dif_pinmux_registers_lock(&dif_pinmux_),
+            kDifPinmuxRegistersLockErrorNone);
+
+  EXPECT_FALSE(dif_pinmux_.write_enabled);
+}
+
+// dif_pinmux_insel_set() tests.
+
+TEST_F(DifPinmuxTest, InselSetFailureInvalidArgumentPinmux) {
+  EXPECT_EQ(dif_pinmux_insel_set(nullptr, kDifPinmuxPeripheralInGpio0,
+                                 kDifPinmuxInselTieToZero),
+            kDifPinmuxInselSetErrorInvalidArgument);
+}
+
+TEST_F(DifPinmuxTest, InselSetFailureBadMioSelect) {
+  DeviceInit();
+
+  int less_than_min = (int)kDifPinmuxInselTieToZero - 1;
```

Comment:
Prefer
```cc
auto foo = static_cast<int>(...);
```

In general, avoid C-like casts like the plague in C++ and use `static_cast` instead.

Reply:
Done.